[2단계 - 장바구니 협업] 콤피(류현승) 미션 제출합니다.

compy-ryu opened the pull request:
안녕하세요! 율리!
콤피 🦖 입니다.

드디어 마지막 리뷰네요! 🥺

마지막 리뷰인만큼, 이번 미션에서는 새로운 기능 추가에 목적을 두는 것 보다는
기존 구조에 대해 다시 한번 생각해보고 전반적으로 리팩토링 해보았습니다!

부족하고 수상한(?) 코드이지만! 이번에도 잘 부탁드립니다!
감사합니다!!!! 👍

<br>


## 실행 방법 🚀
> 기본 MSW 모킹 API로 연결되도록 설정되어있습니다.

- `yarn start`
- 화면 좌측 상단의 API 서버 선택을 통해 백엔드 API 변경이 가능합니다.

![스크린샷 2022-06-09 오후 11 41 17](https://user-images.githubusercontent.com/94832610/172874793-d7d11bdd-29f3-4415-ad16-fdb3c89ec57b.png)



<br>

## 여쭤볼 점 & 시도한 점 🤔
하단 코멘트 목록 참조 부탁드립니다! 🙇‍♂️


<br>

## etc 🎸
* 율리님께서 1단계에서 말해주신 인증/인가 부분도 라우터 측에서 바로 알아보고, 처리할 수 있도록 전체적으로 개선해보았습니다.
  * https://github.com/compy-ryu/react-shopping-cart-prod/blob/compy-ryu-step2/src/Routes.jsx

* 회원 탈퇴 기능은 회원 정보 수정 페이지 하단에 작은 글씨로 숨어있습니다!

![스크린샷 2022-06-09 오후 11 14 27](https://user-images.githubusercontent.com/94832610/172868490-a29dca2d-4988-4468-bcb8-efe09f304829.png)

compy-ryu committed:
♻️ : 비동기 처리 컴포넌트 적용

compy-ryu committed:
✨ : useReduxState 커스텀 훅 추가 및 적용

compy-ryu committed:
💄 : 대체 폰트 지정

compy-ryu committed:
💄 : 웹폰트 CDN에서 Local으로 변경

compy-ryu committed:
💄 : 컬러 상수 네이밍 정리

compy-ryu committed:
♻️ : Styled Component 네이밍 규칙 변경

- "Styled.~" => "S.~"

compy-ryu committed:
♻️ : 페이지 URL 상수 분리 처리

compy-ryu committed:
✨ : 체크박스 컴포넌트 추가

compy-ryu committed:
♻️ : 가격 포멧 변환 부분 유틸으로 분리

compy-ryu committed:
✨ : 카운터 컴포넌트 추가

compy-ryu committed:
♻️ : Px => Rem 사이즈 계산 부분 유틸리티로 분리

compy-ryu committed:
✨ : 아이콘 컴포넌트 추가

compy-ryu committed:
💄 : 테마 상수 업데이트

compy-ryu committed:
💄 : 체크박스 컴포넌트 스타일 수정

compy-ryu committed:
📦 : MSW 설치 및 상품 목록 로드 API 모킹

compy-ryu committed:
🔧 : 개발 환경에서 API 서버 지정 가능하도록 설정 변경

- yarn start => Real Server으로 연결
- yarn start:msw => MSW으로 연결

compy-ryu committed:
✨  : 상품 목록 로드 Mocking API 실제 서버 동작과 동일하게 설정

- 페이지 번호 및 페이지 별 컨텐츠 갯수 지정
- 응답 헤더를 통해 컨텐츠의  총 갯수 반환

compy-ryu committed:
🔧 : babel 설정 커스텀 및 Emotion babel plugin 설치

compy-ryu committed:
🔧 : Storybook Babel Plugin 설정

compy-ryu committed:
🐛 : MSW 사용 시 페이지 새로고침이 반복되는 버그 수정

compy-ryu committed:
🔧 : 개발 모드에서 ClassName을 통해 스타일 위치 파악 가능하도록 설정

compy-ryu committed:
✨ : 페이지 타이틀 컴포넌트 추가

compy-ryu committed:
💄 : Checkbox 스타일 수정

- 체크박스와 텍스트 간격 수정

compy-ryu committed:
🐛 : Checkbox Type 버그 수정

compy-ryu committed:
💄 : Button 컴포넌트 스타일 수정

- 사이즈 지정이 없을 시 상위 요소의 사이즈에 따라 변경되도록 설정
- 버튼 타입에서 아이콘/배경 컬러 변경

compy-ryu committed:
💄 : PageTitle 컴포넌트 스타일 수정

- 사이즈 지정이 없을 시 상위 요소 사이즈를 따라가도록 변경

compy-ryu committed:
🚚 : PageTItle 컴포넌트 Title으로 이름 변경

compy-ryu committed:
💄 : Title 컴포넌트 type에 따라 스타일 변경 되도록 수정

compy-ryu committed:
✨ : FlexContainer 컴포넌트 추가

compy-ryu committed:
✨ : TextUnderline 컴포넌트 추가

- 텍스트에 밑줄 이펙트를 추가합니다.

compy-ryu committed:
💄 : 컬러 팔레트 업데이트

- 브랜드/상태 컬러 추가
- 전체 컬러 추가

compy-ryu committed:
💄 : Button 컴포넌트 상태 컬러 추가

- 버튼 상태 지정 시 브랜드 컬러에 따라 색상이 변경되도록 개선

compy-ryu committed:
💄 : Checkbox 컴포넌트 체크 영역 크기 조절

compy-ryu committed:
💄 : Counter 컴포넌트 최소 너비 제거

compy-ryu committed:
✨  : FlexContainer 컴포넌트 gap Prop 추가

compy-ryu committed:
💄 : Title 컴포넌트 스타일 변경

- content 타입일 때 margin 제거
- padding 조정

compy-ryu committed:
🐛 : 간헐적으로 ToolTip이 다른 앨리먼트에 가려지는 문제 해결

compy-ryu committed:
✨ : ProductItem 방향 지정 기능 추가

- direction prop에 따라 이미지/목록형으로 출력합니다.

compy-ryu committed:
✨ : CartItem 컴포넌트 추가

- 장바구니 목록에서 상품을 출력하여줍니다.

compy-ryu committed:
🐛 : getNumberFormatter 함수 버그 수정

- String 타입이 넘어올 시 오류 출력되는 문제 해결

compy-ryu committed:
💄 : ProductList 페이지 스타일 컴포넌트 네이밍 수정

compy-ryu committed:
✨ : 장바구니 페이지 초안 작성

compy-ryu committed:
✏️ : Counter 컴포넌트 children 타입 변경

compy-ryu committed:
✨ : 장바구니 페이지 라우팅 처리

compy-ryu committed:
♻️ : Checkbox 상태 처리를 위한 값 변경

compy-ryu committed:
✏️ : CartItem, ProductItem 상품 정보 타입 수정

compy-ryu committed:
♻️ : 상품 목록 이미지 영역에서 fill prop 지정

compy-ryu committed:
📦 : immer 패키지 추가

compy-ryu committed:
♻️ : Action Type 이름 변경

compy-ryu committed:
♻️ : 사용하지 않는 코드 제거 및 비동기 상태 생성 함수 이름 변경

compy-ryu committed:
♻️ : redux thunk와 action 함수 분리 및 변경된 액션 타입 적용

compy-ryu committed:
♻️ : products 리듀서 변경된 액션 타입 설정 및 immer 적용

compy-ryu committed:
♻️ : (🙋‍♂️) thunk 호출자에서 비동기 처리 대응

- API 통신 후 return 값 반환을 통해 호출자에서 then  호출 가능하도록 개선
- 리뷰어님에게 여쭤보기

compy-ryu committed:
🚚 : 리덕스 액션 함수 및 Thunk 디렉터리 구조 수정

compy-ryu committed:
✨ : 장바구니 추가 / 제거 / 수정 API 추가

compy-ryu committed:
✨ : 장바구니 추가/불러오기 액션 함수 작성

compy-ryu committed:
♻️ : 비동기 통신 컨텐츠 로드 유무 상태 추가

compy-ryu committed:
♻️ : 장바구니 리듀서 변경된 구조에 맞게 변경

compy-ryu committed:
💄 : Counter Disabled 스타일 지정

compy-ryu committed:
✨ : Counter 컴포넌트 이벤트 핸들러 추가

compy-ryu committed:
✨ : 장바구니 내 상품 수량 및 개별 체크 상태 변경 기능 추가

compy-ryu committed:
✨ : 장바구니 내 상품 제거 기능 추가

compy-ryu committed:
✨ : 장바구니 내 체크 된 상품 선택 삭제 기능 추가

compy-ryu committed:
✨ : 장바구니 상품 목록 전체 선택 및 해제 기능 추가

compy-ryu committed:
✨ : 장바구니 결제 예상 금액 및 선택 수량 출력 기능 추가

compy-ryu committed:
♻️ : useCart 커스텀훅 사용 방식 변경

compy-ryu committed:
♻️ : 상품 목록 페이지별 불러오는 상품 수량 수정

compy-ryu committed:
♻️ : children 요소 prop 타입 제거

compy-ryu committed:
✨ : 상품 목록이 장바구니에 존재할 시 토글 처리

compy-ryu committed:
🐛 : 장바구니에 상품 추가 시 상태가 갱신되지 않던 버그 해결

compy-ryu committed:
✨ : 로고 및 장바구니 라우터 연결

compy-ryu committed:
✨ : Header 컴포넌트 장바구니 수량 확인 기능 추가

compy-ryu committed:
✨ : 장바구니 영역 Mocking API 구현

compy-ryu committed:
🐛 : 장바구니 페이지에서 MSW 작동 되지 않던 문제 해결

- Pathname으로 인한 버그로 확인

compy-ryu committed:
🔧 : 환경 변수 설정 정리

compy-ryu committed:
⚰️ : 사용하지 않는 코드 제거

compy-ryu committed:
✅ : 장바구니 리듀서 상태 CURD 테스트 코드 추가

compy-ryu committed:
♻️ : Button 컴포넌트 PropType 정리 및 스토리북 예시 추가

compy-ryu committed:
💄 : Checkbox 컴포넌트 스토리북 예시 추가

compy-ryu committed:
💄 : Counter 컴포넌트 스토리북 기본 값 지정

compy-ryu committed:
♻️ : FlexContainer 컴포넌트 Import 개행 정리

compy-ryu committed:
⚰️ : TextUnderline 컴포넌트 사용하지 않는 Import 제거

compy-ryu committed:
💄 : Title 컴포넌트 스토리북 작성

compy-ryu committed:
💄 : 장바구니 상품 목록 스토리북 정리

- 이벤트 핸들러 기본 값 지정
- 누락된 타입 지정

compy-ryu committed:
♻️ : 장바구니 제거 시 alert 메시지 수정

compy-ryu committed:
♻️ : 상품 목록 페이지 장바구니 추가/제거 시 alert 메시지 수정

compy-ryu committed:
✨ : StatusMessage 컨텐츠가 비어있을 때 상태 추가

compy-ryu committed:
✨ : 장바구니가 비어 있을 때 상태 메시지 추가

compy-ryu committed:
💄 : 장바구니 상품 주문 영역 화면 출력 높이 조정

compy-ryu committed:
♻️ : 범용 컴포넌트 Import 정리

compy-ryu committed:
🤡 : mocking API용 데이터 업데이트

compy-ryu committed:
📝 : 기능 요구사항 업데이트

compy-ryu committed:
♻️ : mockServiceWorker Prettier 적용

compy-ryu committed:
♻️ : Counter 컴포넌트 Prop 전달 방식 변경

compy-ryu committed:
♻️ : ProductItem 컴포넌트 Switch문 제거

compy-ryu committed:
🐛 : API 비동기 상태 처리 버그 수정

- isLoaded 처리 초기화 처리 추가

compy-ryu committed:
♻️ : useReduxState 커스텀 훅 제거

compy-ryu committed:
♻️ : 상품 관련 상태 정보 dispatch 사용성 개선

- action 및 thunk를 가져올 때 일괄적으로 가져온 후 컴포넌트에서 Dispatch 하도록 가독성 개선

compy-ryu committed:
♻️ : 장바구니 관련 상태 정보 dispatch 사용성 개선

- action 및 thunk를 가져올 때 일괄적으로 가져온 후 컴포넌트에서 Dispatch 하도록 가독성 개선
- action 함수 네이밍 통일

compy-ryu committed:
⚰️ : 사용하지 않는 Import 제거

compy-ryu committed:
♻️ : 장바구니 Action 및 Thunk 영역 useCart 커스텀 훅으로 이동

compy-ryu committed:
🐛 : 장바구니 목록 API가 중복 호출되는 버그 수정

compy-ryu committed:
♻️ : 장바구니 페이지 컨테이너 분리

- 코드 가독성을 위해 컨테이너 단위로 추가 분리 처리

compy-ryu committed:
♻️ : 장바구니 주문하기 컨테이너 분리 및 장바구니 내 컨테이너 네이밍 변경

compy-ryu committed:
♻️ : 리듀서에서 비동기 상태 컨텐츠와 분리

compy-ryu committed:
♻️ : 장바구니 Reducer, immer 라이브러리 적용

compy-ryu committed:
♻️ : 상품 정보 reducer, immer 라이브러리 적용

- 일관성을 위해 모두 적용하였습니다.

compy-ryu committed:
♻️ : 장바구니 상품 업데이트/제거 시 오류 처리

compy-ryu committed:
♻️ : 상품 목록에서 장바구니 추가/제거 시 실패 시에만 경고창이 출력되도록 변경

compy-ryu committed:
🔧 : prettier import 자동 정렬 설정 및 적용

compy-ryu committed:
[2단계 - 장바구니] 콤피(류현승) 미션 제출합니다. (#131)

* 📦 : 프로젝트 세팅

Co-authored-by: Usage <kyr9389@naver.com>

* 📝 : 기능 목록 초안 작성

Co-authored-by: Usage <kyr9389@naver.com>

* 📦 : ESLint, Prettier 패키지 설정

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : FontAwesome 설치

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : 글로벌 스타일 설정

* ✨ : Header, Footer, Layout 컴포넌트 추가

* ✨ : IconButton 컴포넌트 추가

* ✨ : 상품 목록 내 요소 컴포넌트 추가

* refactor: 아이콘 코드 상수 분리

* 🔧 : .gitignore 설정

* 💄 : root 폰트 사이즈 지정

* 💄 : Footer 태그 수정

* 💄 : reset css 적용

* ♻️ : IconButton 컴포넌트 화살표 함수로 변경

* 🔧 : ESLint 예외규칙 추가

* ✨ : Header 컴포넌트 추가

* ✨ : Footer 컴포넌트 추가

* 💄  : 상품 목록 내 요소 컴포넌트 스타일 추가

* 💄 : Layout 컴포넌트 스타일 작성

* 🚚 : 컴포넌트 Storybook 위치 변경

* 💄  : ProductList 페이지 스타일 추가

* 💄 : 글로벌 스타일 및 테마 추가

* 📦 : Redux 관련 패키지 설치

* 📦 : React logger 패키지 설치

Co-authored-by: Usage <kyr9389@naver.com>

* 🔧 : ESLint 예외 규칙 추가

* ✨ : 상품 목록 API 통신 추가

Co-authored-by: Usage <kyr9389@naver.com>

* ✨ : 리덕스 상태 관리 추가

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : 화면 너비가 좁을 시 상품 목록이 깨지는 문제 해결

Co-authored-by: Usage <kyr9389@naver.com>

* ♻️ : API 통신 결과 상태 상수 분리

Co-authored-by: Usage <kyr9389@naver.com>

* ✨ : 상품 목록 API를 통해 상품 출력 기능 추가

Co-authored-by: Usage <kyr9389@naver.com>

* 🚚 : 액션 타입 목록 파일명 수정

Co-authored-by: Usage <kyr9389@naver.com>

* ✨ : 장바구니 담기 기능 구현

Co-authored-by: Usage <kyr9389@naver.com>

* 📝 : 기능 목록 업데이트

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : Header 로고 추가

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : IconButton Mouse Hover 효과 추가

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : Footer 스타일 작성

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : 화면 최소 너비 지정

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : 상품 사진 확대 효과 추가

Co-authored-by: Usage <kyr9389@naver.com>

* 🔧 : 환경 변수 설정

Co-authored-by: Usage <kyr9389@naver.com>

* ♻️ : 장바구니 담기 시 알림 메시지 추가

Co-authored-by: Usage <kyr9389@naver.com>

* ♻️ : 장바구니 추가 액션 변수명 수정

Co-authored-by: Usage <kyr9389@naver.com>

* ♻️ : 서버 통신 시 오류 처리 기능 추가

Co-authored-by: Usage <kyr9389@naver.com>

* 💄 : 스타일 색상 상수 분리

Co-authored-by: Usage <kyr9389@naver.com>

* 📦 : Github Pages 패키지 설치

Co-authored-by: Usage <kyr9389@naver.com>

* 🔧 : 환경 변수 내 API 서버 주소 변경

* ♻️ : API 구조 및 리듀서 처리 방식 변경

- 로딩, 성공, 실패 처리 상태 추가
- API 통신 시 상태 관리 일관성 유지되도록 구조 개선
- 리듀서 코드 라인 개선
- 액션 코드 가독성 개선
- API 도메인 분리

* ✏️  : API Mock Data 변경 대응

* 💄 : 상품 목록 타이틀이 길어질 때 레이아웃이 깨지는 문제 해결

* 🔧 : Prettier EOF 설정 수정

* 📦 : Storybook 정식 버전 변경

* ♻️ : 컴포넌트 선언 방식 수정

* 💄 : 폰트 추가

- 기본 폰트 나눔바른고딕 적용
- 일부 요소에서 기랑해랑체 적용

* 💄 : 애니메이션 스타일 유틸 추가

* 🚚 : 로고 이미지 위치 수정

* ✨ : 초기 접속 시 사전 로드할 이미지 설정

* 📦 : 사용하지 않는 패키지 파일 제거

* ✨ : 로딩/오류 메시지 컴포넌트 추가

* ♻️ : 웹폰트 사전 로드 설정

* ♻️ : 화살표 함수를 통한 컴포넌트 선언 제거

* ✨ : 비동기 상태 분기 커스텀 훅 추가

- 로딩, 오류, 완료 상태에 따라 렌더링 될 요소가 지정

* 💄 : 상품 목록 컴포넌트 스타일 작성 방식 변경

* 💄 : 스타일 변수 네이밍 수정

* 🔧 : Storybook/Emotion 충돌 이슈 해결

- https://xo.dev/fix-storybook-emotion-11-error/

* 💄 : Storybook 글로벌 스타일 적용

* ♻️ : IconButton 컴포넌트 Button 컴포넌트로 통합

* ⚰️ : 임시 주석 제거

* 💄 : 컬러 상수 네이밍 규칙 변경

* 💄 : border-radius 값 상수 처리

* 💄 : Button 컴포넌트 컬러 상수 처리

* 💄 : Button 컴포넌트 스토리북 타이틀 수정

* ✨ : ToolTip 컴포넌트 추가

* ✨ : 장바구니 담기 시 툴팁이 출력되도록 수정

* ♻️ : Header 컴포넌트 스타일링 방식 변경

* ♻️ : 장바구니 추가 핸들러 페이지 컴포넌트로 이동

* 🚚 : 상품 목록 페이지 컴포넌트 위치 변경

* 💄 : StatusMessage 컴포넌트 스토리북 추가

* 💄 : 스토리북 리덕스 적용

* 💄 : 기본 레이아웃 컴포넌트 스타일 및 스토리북 추가

* 🚚 : 스토리북 템플릿 위치 수정

* ➕ : React Router 패키지 설치

* ✨ : Not Found 페이지 추가

* 💄 : Button 컴포넌트 아이콘 사이즈 조정

* ♻️ : 레이아웃 컴포넌트 React Router Outlet 적용

* ♻️ : 아이콘 코드 상수 추가

* ✨ : React Router 적용

- 화면 및 products 페이지 상품 목록 페이지 연결
- 없는페이지 Not Found 페이지로 연결

* ♻️ : Symbol 타입 제거

* ✨ : API 통신 시 타임아웃 처리

* 🔧 : ESLint 예외 규칙 추가

* ✨ : 상품 목록 로드 액션, 페이지 설정 인자 추가

* ⚰️ : 사용하지 않는 import 제거

* 🔧 : package.json 홈페이지 설정

* 🐛 : 배포 시 라우터 디렉토리 경로 문제 해결

* ✨ : 비동기 상태에 따른 분기 처리용 컴포넌트 구현

* ⚰️ : 사용하지 않는 코드 제거

* 🚚 : 디렉터리명 수정

- IfAsync => SwitchAsync

* ♻️ : 비동기 처리 컴포넌트 적용

* ✨ : useReduxState 커스텀 훅 추가 및 적용

* 💄 : 대체 폰트 지정

* 💄 : 웹폰트 CDN에서 Local으로 변경

* 💄 : 컬러 상수 네이밍 정리

* ♻️ : Styled Component 네이밍 규칙 변경

- "Styled.~" => "S.~"

* ♻️ : 페이지 URL 상수 분리 처리

* ✨ : 체크박스 컴포넌트 추가

* ♻️ : 가격 포멧 변환 부분 유틸으로 분리

* ✨ : 카운터 컴포넌트 추가

* ♻️ : Px => Rem 사이즈 계산 부분 유틸리티로 분리

* ✨ : 아이콘 컴포넌트 추가

* 💄 : 테마 상수 업데이트

* 💄 : 체크박스 컴포넌트 스타일 수정

* 📦 : MSW 설치 및 상품 목록 로드 API 모킹

* 🔧 : 개발 환경에서 API 서버 지정 가능하도록 설정 변경

- yarn start => Real Server으로 연결
- yarn start:msw => MSW으로 연결

* ✨  : 상품 목록 로드 Mocking API 실제 서버 동작과 동일하게 설정

- 페이지 번호 및 페이지 별 컨텐츠 갯수 지정
- 응답 헤더를 통해 컨텐츠의  총 갯수 반환

* 🔧 : babel 설정 커스텀 및 Emotion babel plugin 설치

* 🔧 : Storybook Babel Plugin 설정

* 🐛 : MSW 사용 시 페이지 새로고침이 반복되는 버그 수정

* 🔧 : 개발 모드에서 ClassName을 통해 스타일 위치 파악 가능하도록 설정

* ✨ : 페이지 타이틀 컴포넌트 추가

* 💄 : Checkbox 스타일 수정

- 체크박스와 텍스트 간격 수정

* 🐛 : Checkbox Type 버그 수정

* 💄 : Button 컴포넌트 스타일 수정

- 사이즈 지정이 없을 시 상위 요소의 사이즈에 따라 변경되도록 설정
- 버튼 타입에서 아이콘/배경 컬러 변경

* 💄 : PageTitle 컴포넌트 스타일 수정

- 사이즈 지정이 없을 시 상위 요소 사이즈를 따라가도록 변경

* 🚚 : PageTItle 컴포넌트 Title으로 이름 변경

* 💄 : Title 컴포넌트 type에 따라 스타일 변경 되도록 수정

* ✨ : FlexContainer 컴포넌트 추가

* ✨ : TextUnderline 컴포넌트 추가

- 텍스트에 밑줄 이펙트를 추가합니다.

* 💄 : 컬러 팔레트 업데이트

- 브랜드/상태 컬러 추가
- 전체 컬러 추가

* 💄 : Button 컴포넌트 상태 컬러 추가

- 버튼 상태 지정 시 브랜드 컬러에 따라 색상이 변경되도록 개선

* 💄 : Checkbox 컴포넌트 체크 영역 크기 조절

* 💄 : Counter 컴포넌트 최소 너비 제거

* ✨  : FlexContainer 컴포넌트 gap Prop 추가

* 💄 : Title 컴포넌트 스타일 변경

- content 타입일 때 margin 제거
- padding 조정

* 🐛 : 간헐적으로 ToolTip이 다른 앨리먼트에 가려지는 문제 해결

* ✨ : ProductItem 방향 지정 기능 추가

- direction prop에 따라 이미지/목록형으로 출력합니다.

* ✨ : CartItem 컴포넌트 추가

- 장바구니 목록에서 상품을 출력하여줍니다.

* 🐛 : getNumberFormatter 함수 버그 수정

- String 타입이 넘어올 시 오류 출력되는 문제 해결

* 💄 : ProductList 페이지 스타일 컴포넌트 네이밍 수정

* ✨ : 장바구니 페이지 초안 작성

* ✏️ : Counter 컴포넌트 children 타입 변경

* ✨ : 장바구니 페이지 라우팅 처리

* ♻️ : Checkbox 상태 처리를 위한 값 변경

* ✏️ : CartItem, ProductItem 상품 정보 타입 수정

* ♻️ : 상품 목록 이미지 영역에서 fill prop 지정

* 📦 : immer 패키지 추가

* ♻️ : Action Type 이름 변경

* ♻️ : 사용하지 않는 코드 제거 및 비동기 상태 생성 함수 이름 변경

* ♻️ : redux thunk와 action 함수 분리 및 변경된 액션 타입 적용

* ♻️ : products 리듀서 변경된 액션 타입 설정 및 immer 적용

* ♻️ : (🙋‍♂️) thunk 호출자에서 비동기 처리 대응

- API 통신 후 return 값 반환을 통해 호출자에서 then  호출 가능하도록 개선
- 리뷰어님에게 여쭤보기

* 🚚 : 리덕스 액션 함수 및 Thunk 디렉터리 구조 수정

* ✨ : 장바구니 추가 / 제거 / 수정 API 추가

* ✨ : 장바구니 추가/불러오기 액션 함수 작성

* ♻️ : 비동기 통신 컨텐츠 로드 유무 상태 추가

* ♻️ : 장바구니 리듀서 변경된 구조에 맞게 변경

* 💄 : Counter Disabled 스타일 지정

* ✨ : Counter 컴포넌트 이벤트 핸들러 추가

* ✨ : 장바구니 내 상품 수량 및 개별 체크 상태 변경 기능 추가

* ✨ : 장바구니 내 상품 제거 기능 추가

* ✨ : 장바구니 내 체크 된 상품 선택 삭제 기능 추가

* ✨ : 장바구니 상품 목록 전체 선택 및 해제 기능 추가

* ✨ : 장바구니 결제 예상 금액 및 선택 수량 출력 기능 추가

* ♻️ : useCart 커스텀훅 사용 방식 변경

* ♻️ : 상품 목록 페이지별 불러오는 상품 수량 수정

* ♻️ : children 요소 prop 타입 제거

* ✨ : 상품 목록이 장바구니에 존재할 시 토글 처리

* 🐛 : 장바구니에 상품 추가 시 상태가 갱신되지 않던 버그 해결

* ✨ : 로고 및 장바구니 라우터 연결

* ✨ : Header 컴포넌트 장바구니 수량 확인 기능 추가

* ✨ : 장바구니 영역 Mocking API 구현

* 🐛 : 장바구니 페이지에서 MSW 작동 되지 않던 문제 해결

- Pathname으로 인한 버그로 확인

* 🔧 : 환경 변수 설정 정리

* ⚰️ : 사용하지 않는 코드 제거

* ✅ : 장바구니 리듀서 상태 CURD 테스트 코드 추가

* ♻️ : Button 컴포넌트 PropType 정리 및 스토리북 예시 추가

* 💄 : Checkbox 컴포넌트 스토리북 예시 추가

* 💄 : Counter 컴포넌트 스토리북 기본 값 지정

* ♻️ : FlexContainer 컴포넌트 Import 개행 정리

* ⚰️ : TextUnderline 컴포넌트 사용하지 않는 Import 제거

* 💄 : Title 컴포넌트 스토리북 작성

* 💄 : 장바구니 상품 목록 스토리북 정리

- 이벤트 핸들러 기본 값 지정
- 누락된 타입 지정

* ♻️ : 장바구니 제거 시 alert 메시지 수정

* ♻️ : 상품 목록 페이지 장바구니 추가/제거 시 alert 메시지 수정

* ✨ : StatusMessage 컨텐츠가 비어있을 때 상태 추가

* ✨ : 장바구니가 비어 있을 때 상태 메시지 추가

* 💄 : 장바구니 상품 주문 영역 화면 출력 높이 조정

* ♻️ : 범용 컴포넌트 Import 정리

* 🤡 : mocking API용 데이터 업데이트

* 📝 : 기능 요구사항 업데이트

* ♻️ : mockServiceWorker Prettier 적용

* ♻️ : Counter 컴포넌트 Prop 전달 방식 변경

* ♻️ : ProductItem 컴포넌트 Switch문 제거

* 🐛 : API 비동기 상태 처리 버그 수정

- isLoaded 처리 초기화 처리 추가

* ♻️ : useReduxState 커스텀 훅 제거

* ♻️ : 상품 관련 상태 정보 dispatch 사용성 개선

- action 및 thunk를 가져올 때 일괄적으로 가져온 후 컴포넌트에서 Dispatch 하도록 가독성 개선

* ♻️ : 장바구니 관련 상태 정보 dispatch 사용성 개선

- action 및 thunk를 가져올 때 일괄적으로 가져온 후 컴포넌트에서 Dispatch 하도록 가독성 개선
- action 함수 네이밍 통일

* ⚰️ : 사용하지 않는 Import 제거

* ♻️ : 장바구니 Action 및 Thunk 영역 useCart 커스텀 훅으로 이동

* 🐛 : 장바구니 목록 API가 중복 호출되는 버그 수정

* ♻️ : 장바구니 페이지 컨테이너 분리

- 코드 가독성을 위해 컨테이너 단위로 추가 분리 처리

* ♻️ : 장바구니 주문하기 컨테이너 분리 및 장바구니 내 컨테이너 네이밍 변경

* ♻️ : 리듀서에서 비동기 상태 컨텐츠와 분리

* ♻️ : 장바구니 Reducer, immer 라이브러리 적용

* ♻️ : 상품 정보 reducer, immer 라이브러리 적용

- 일관성을 위해 모두 적용하였습니다.

* ♻️ : 장바구니 상품 업데이트/제거 시 오류 처리

* ♻️ : 상품 목록에서 장바구니 추가/제거 시 실패 시에만 경고창이 출력되도록 변경

* 🔧 : prettier import 자동 정렬 설정 및 적용

Co-authored-by: Usage <kyr9389@naver.com>

compy-ryu committed:
Merge branch 'compy-ryu' of https://github.com/woowacourse/react-shopping-cart into compy-ryu-step2

compy-ryu committed:
🐛  : 변경된 비동기 상태가 적용되지 않아, API가 중복 호출되는 문제 해결

compy-ryu committed:
♻️ : reselect를 활용한 체크 목록 종북 계산 방지

compy-ryu committed:
✨ : InputField 컴포넌트 추가

Co-authored-by: kyoul <sion11290805@gmail.com>

compy-ryu committed:
✨ : FieldSet 컴포넌트 추가

Co-authored-by: kyoul <sion11290805@gmail.com>

compy-ryu committed:
🚚 : 페이지 컴포넌트 네이밍 규칙 수정

- 페이지 컴포넌트는 컴포넌트 이름 뒤에 Page를 붙인다.

Co-authored-by: kyoul <sion11290805@gmail.com>

compy-ryu committed:
💄 : FieldSet 컴포넌트 스타일 변경

- Input 컨테이너 FlexContainer 적용

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
💄 : InputFIeld 컴포넌트 스타일 변경

- padding 수정

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
🚚 : 레이아웃 이름 및 디렉토리 위치 변경

- Layout => ShoppingLayout

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 페이지에 따른 레이아웃 분기 처리

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
💄 : 글로벌 스타일 화면 높이 지정

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
⚰️ : Routes 사용하지 않는 코드 및 주석 제거

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
🔧 : ESLint 예외 규칙 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 서비스 레이아웃 컴포넌트 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원가입 페이지 컴포넌트 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
♻️ : Button 컴포넌트 Prop 이름 변경

- state => status

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 로그인 페이지 컴포넌트 추가 및 라우팅 처리

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
💄 : 회원가입 페이지 버튼 스타일 변경

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 유효성 검사 유틸리티 함수 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 관련 API 모킹 처리

- 가입, 로그인, 중복체크, 회원 정보 조회 API 모킹

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : Button 컴포넌트 비활성 상태 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
💄 : InputField 컴포넌트 hover 버그 수정

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
♻️  : Button 컴포넌트 prop 명칭 변경 대응

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 유효성 검사 유틸리티 함수 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 엘리먼트 관련 유틸리티 함수 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
♻️ : Thunk를 통한 dispatch 방식 구조 개선

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
♻️ : 범용 컴포넌트 로드 처리

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 관련 액션 타입 상수 선언

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
🔧 : ESLint 예외규칙 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
⚰️ : useCart 콘솔 로그 제거

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 관련 Redux thunk 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 관련 액션 함수 선언

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 관련 API 명세 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : Request 유틸리티 내 엑세스 토큰 전송 옵션 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 상수 업데이트

- 페이지 리스트
- 액세스 토큰 쿠키 이름 상수화

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : useFormValidation 커스텀 훅 기능 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : useUserSession 커스텀 훅 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 쿠키 관련 유틸리티 함수 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 유효성 검사 관련 유틸리티 업데이트

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : MSW 회원 정보 관련 모킹 API 명세 최신화

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 리듀서 추가

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원가입 페이지 기능 구현

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 로그인 페이지 기능 구현

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 로그아웃 페이지 기능 구현

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 유저 정보 세션 연결

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 관리 페이지 초안 작성

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
📝 : 기능 목록 문서 작성

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
✨ : 회원 정보 관리 및 로그아웃 페이지 라우팅 처리

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
♻️ : Header 컴포넌트 로그인, 로그아웃 버튼 기능 구현

Co-authored-by: kyoul <yunjin-kim@users.noreply.github.com>

compy-ryu committed:
Merge commit 'a9cfe49b29801e8d96a54f8dca9a27921badcad0' into compy-ryu-step1

compy-ryu committed:
♻️ : getFormData 함수 Form API로 변경

compy-ryu committed:
✨ : 리덕스 구조 정리를 위한 redux-template 기능 추가

compy-ryu committed:
♻️ : createAsyncState 함수 위치 변경

compy-ryu committed:
♻️ : 리덕스 상품 영역 상태 변경된 구조 적용

compy-ryu committed:
♻️ : 리덕스 상품 영역 상태 네이밍 정리

compy-ryu committed:
♻️ : 리덕스 액션 함수에서 구조 분해 제거

compy-ryu committed:
♻️ : 리덕스 장바구니 영역 상태 변경된 구조로 적용

compy-ryu committed:
♻️ : 리덕스 장바구니 영역 상태 사용 중인 페이지 컴포넌트 구조 변경 사항 적용

compy-ryu committed:
🐛 : 장바구니 아이템 업데이트/제거 시 성공 상태로 갱신되지 않는 버그 수정

compy-ryu committed:
🐛 : 장바구니 Thunk 적용

compy-ryu committed:
♻️ : 리덕스 회원 정보 영역 상태 변경된 구조 적용

compy-ryu committed:
♻️ : 회원 정보 상태 변경된 구조에 맞추어 컴포넌트 구조 개선

compy-ryu committed:
♻️ : store 위치 변경

compy-ryu committed:
⚰️ : 사용하지 않는 코드 제거

compy-ryu committed:
Merge branch 'compy-ryu' of https://github.com/woowacourse/react-shopping-cart-prod into compy-ryu-step2-dev

compy-ryu committed:
🚑 : 컨플릭트 해결

compy-ryu committed:
♻️ : 회원 정보 API 위치 개행 정리

compy-ryu committed:
🐛 : 액세스 토큰 만료 체크가 페이지 로드 후 한번만 이뤄지는 문제 해결

compy-ryu committed:
🐛 : 닉네임 중복 체크가 정상적으로 이뤄지지 않는 문제 해결

compy-ryu committed:
♻️ : MSW 회원 정보 영역 로그인 실패 메시지 수정 및 중복 정보 체크 분기 처리

compy-ryu committed:
♻️ : requestUtils 오류 처리 개선 및 리팩터링

- 서버 통신 실패 시 http 상태 번호 반환
- 액세스 토큰 추가 부분 함수 분리
- 서버 응답 중 Body 값이 비어있을 때 무조건 실패로 처리되는 버그 수정

compy-ryu committed:
♻️ : request 함수 백엔드와 오류 처리 방식 통일

compy-ryu committed:
♻️ : 유저 정보 API 명세 응답 값 구조분해 처리

- 의도 : API 명세를 계속해서 확인하는 작업을 줄이기 위해 응답 값을 바로 알 수 있도록 구조 분해 처리

compy-ryu committed:
♻️ : API 명세 영역 응답 값 구조분해 제거

compy-ryu committed:
♻️ : request 함수 응답 방식 변경에 따른 수정

compy-ryu committed:
✨ : useDispatchEvent 추가

- dispatch 호출 영역에서 업데이트 이후 상태 로드를 위한 커스텀 훅

compy-ryu committed:
✨ : useFetch 커스텀 훅 추가

- 컴포넌트 내에서 API 통신 시 useFetch를 통해 상태를 제어합니다.

compy-ryu committed:
♻️ : 회원가입 API 통신 중 버튼이 비활성화 되도록 구조 개선

compy-ryu committed:
🔧 : Package 홈페이지 설정 변경

compy-ryu committed:
🔧 : 백엔드 크루용 env 환경 변수 추가

compy-ryu committed:
♻️ :useFormValidation으로 이벤트 핸들러 위치 변경

compy-ryu committed:
✨ : API 명세 최신화

compy-ryu committed:
✨ : Button 컴포넌트 너비 설정 기능 추가

compy-ryu committed:
✨ : InputField 기본 값 및 비활성화 옵션 추가

compy-ryu committed:
💄 : FieldSet 컴포넌트 설명 텍스트 margin 지정

compy-ryu committed:
♻️ : reqest 함수 옵션 지정 방식 수정

compy-ryu committed:
⚰️ : 사용하지 않는 코드 제거

compy-ryu committed:
♻️ : 모킹용 DB 백엔드와 통일

compy-ryu committed:
✨ : 회원 정보 수정 및 패스워드 확인/변경 API 모킹

compy-ryu committed:
♻️ : 상품 목록 모킹 API 최신화

compy-ryu committed:
🐛 : 상품 목록 API 응답 값 백엔드와 통일

compy-ryu committed:
✨ : 회원 정보 수정 및 패스워드 변경 기능 추가

compy-ryu committed:
♻️ : Checkbox 컴포넌트 비제어 컴포넌트에서도 사용 가능하도록 개선

compy-ryu committed:
♻️ : 변경된 장바구니 API 스펙 추가

compy-ryu committed:
♻️ : 장바구니 추가 시 변경된 API 스펙 적용

compy-ryu committed:
♻️ : 장바구니 영역 백엔드 API 스펙과 동일하게 변경

compy-ryu committed:
🐛 : 장바구니 추가 실패 시 오류 처리

compy-ryu committed:
💄 : FieldSet 컴포넌트 스타일 수정

- 설명에서 화살표 추가

compy-ryu committed:
🚚 : useUserSession 커스텀 훅 이름 변경

- useAuth으로 변경

compy-ryu committed:
🐛 : redux-template 라이브러리 createAsyncState 상태 값 수정

compy-ryu committed:
✨ : 액세스 거부 시 안내 페이지 추가

compy-ryu committed:
♻️ : 로그아웃 기능 상태 처리 구조 개선

compy-ryu committed:
✨ : 로그인 이후 이전 페이지 이동 기능 구현

compy-ryu committed:
🐛 : 프로필 편집 후 유저 정보 갱신되지 않는 버그 수정

compy-ryu committed:
✨ : 회원 탈퇴 기능 구현

compy-ryu committed:
✨ : RequireAuth 컴포넌트 추가 및 페이지별 권한 설정

compy-ryu committed:
💄 : NotFound 페이지 컴포넌트 스타일 수정

compy-ryu committed:
🐛 : CheckBox PropType 오류 해결

compy-ryu committed:
🔧 : ESLint 예외 규칙

compy-ryu committed:
🔧 : 백엔드 API 서버 URL 최신화

compy-ryu committed:
♻️ : 장바구니 페이지 비로그인 상태에서 접근 차단

compy-ryu committed:
✨ : PopupContainer 컴포넌트 추가

compy-ryu committed:
✨  : Backend API 선택 기능 추가

compy-ryu committed:
✨ : ProfileMenu 컴포넌트 추가

compy-ryu committed:
♻️ : MembersLayout 헤더 내 로고를 누를 시 메인 화면으로 이동되도록 개선

compy-ryu committed:
♻️ : ShoppingLayout 헤더 내 회원 전용 메뉴 추가

compy-ryu committed:
♻️ : requestUtils 백엔드 API를 바라보도록 수정

compy-ryu committed:
♻️ : 회원 로그인 처리 구조 개선

compy-ryu committed:
♻️ : 회원 로그아웃 처리 Thunk로 이전

compy-ryu committed:
💄 : 회원 정보 수정 페이지 내 회원 탈퇴 페이지 연결

compy-ryu committed:
♻️ : 장바구니 목록 초기화 액션 추가

compy-ryu committed:
♻️ : 스타일 및 아이콘 상수 추가

compy-ryu committed:
📝 : 기능 구현사항 업데이트

compy-ryu committed:
♻️ : MSW API 명세 최신화

compy-ryu committed:
📦 : MSW 서버 실행 명령어 변경

compy-ryu commented on src/store/product/reducer.js:
```diff
@@ -0,0 +1,25 @@
+import { createReducer, createAsyncState } from 'lib/redux-template';
+
+const initialState = {
+  productList: [],
+  listAsyncState: createAsyncState.initial(),
+};
+
+const reducer = {
+  updateProductList(state, { productList = [] }) {
+    state.productList = productList;
+    state.listAsyncState = createAsyncState.success();
+  },
+
+  updateProductList_Pending(state) {
+    state.listAsyncState = createAsyncState.pending();
+  },
+
+  updateProductList_Error(state, { errorMessage = '' }) {
+    state.listAsyncState = createAsyncState.error(errorMessage);
+  },
+};
+
+const productReducer = createReducer(reducer, initialState);
```
## 시도 1번
### 리덕스와 원만한 합의 시도 🤝

_먼저 저는 리덕스가 마음에 들지 않습니다!_ 🤯

리덕스로 상태를 관리하기 위해 리듀서를 선언하고, 액션 타입을 또 선언하고
거기에서 또 호출할 때 편의를 위해서 액션 크리에이터 함수까지 만들고…

이런 구조로 작성 시 데이터 흐름이 단방향이라 디버깅하기 쉽고, 예측이 가능하여 디버깅이 용이하다!라고 하지만

리듀서가 늘어날 수록 가독성도 정말 좋지 않은데다, 작성하는 코드량도 늘어나고
액션 타입까지 상수로 만드는 행위 자체가 휴먼 에러를 방지한다고도 하지만,  얻는 이점에 비해 불편함과 단점이 너무 크다 생각했습니다.

그리하여, 이번 레벨 2에서 리덕스를 배우고나서 리듀서를 3번정도 구조를 변경해보았는데요.

처음에는 단순 사용법을 따라가고, 두번째는 포코 몰래 immer 쓰다 붙잡히기도 하고
하라는 리덕스는 안 하고 다른 상태 관리 라이브러리 구경도 하고 했었지만!

저한테는 리덕스가 가장 불편해보이더라구요.

그래서 조금이라도 가독성을 개선해보고 싶었고, JS에서도 IDE 자동완성이 알차게 되도록 하기위해

[/lib/redux-template](https://github.com/compy-ryu/react-shopping-cart-prod/blob/compy-ryu-step2/src/lib/redux-template/index.js)에서 createReducer를 작성하여
현재 제가 작성한 리듀서 코드를 리덕스 리듀서에서도 작동될 수 있도록 구조를 수정해보았습니다.

또한, 액션 타입 선언도 너무나도 번거롭고 불편했기에
아래와 같이 createAction을 통해 액션 크리에이터를 바로 만들어주고, Logger에서도 어떤 액션이 실행되는지 확인할 수 있게 구현해보았습니다.

```javascript
const createAction = (action, payload) => ({ type: action.name, payload });

const reducer = {
  updateProductList(state, { productList = [] }) {
    // 생략
  },
}

console.log(createAction(reducer.updateProductList, {productList: []}
// 출력 결과 : { type: updateProductList, payload: { productList : [] } };
```

 
### 하지만 ⚠️
이렇게 구조를 약간씩 변경하였을 때, 생기는 고민이 있습니다.

저 편하자고 툴킷과 같은 공식 라이브러리도 아닌데, 제공해주는 구조를 조금씩 변경한다면
오히려 다른 개발자가 처음 보았을 때 이해하기가 힘들어지지 않을까? 란 생각도 들기도 합니다.

이런 시도는 가능한 지양하고 라이브러리의 원래 사용법을 가능한 유지하는 것이 좋을까요?
아니면 이렇게 편의성을 높이려고 시도하는 부분들은 괜찮을까요?
(단순히 리덕스뿐만 아닌 이외 라이브러리들도 포함입니다!)

율리님은 어떻게 생각하시나요!?

glassyi replied:
원래 사용법에서 편리하게 사용하기 위해 여러 패턴들이 또 등장 하는 것 같아요. 혼자 쓰는 코드가 아니라면 보편적으로 많이 쓰는 패턴으로 적용해서 써야 다른 동료분들이 봤을 때도 쉽게 읽고 접근할 수 있지 않을까 해요. 🙂

compy-ryu replied:
현재 리듀서가 읽기는 더 좋다 생각해서 일단 고치지 않고 그대로 두었습니다!
율리님이 생각하시기에 _아, 이거 너무 수상한데_ 싶으시다면! 개선해보도록 하겠습니다!

compy-ryu commented on src/pages/CartListPage/Containers/CartItemList/index.jsx:
```diff
@@ -1,38 +1,41 @@
-
-    cartCurdAsyncState.isLoaded === false &&
-      alert('서버 오류로 인해 상품 정보 갱신에 실패하였습니다.');
+    dispatchEvent({
+      action: cartThunk.updateItem(id, { quantity }),
+      onStateUpdated: ({ cart }) => {
+        !cart.curdAsyncState.isLoaded && alert('서버 오류로 인해 상품 정보 갱신에 실패하였습니다.');
+      },
+    });
```
## 시도 2번
### 호출 영역에서 리덕스의 최신 상태 호출 🚀

리덕스를 처음 사용했을 때
상태의 변경에 따라 호출 영역(이벤트 핸들러) 바깥에서 useEffect를 추가하고... 감싸고
해당 상태의 변동에 따라 실행 처리해주는 것에서 가장 불편함을 느꼈었는데요!

개인적으로 가독성도 해친다 생각했고
해당 상태를 변경하는 액션들이 늘어난다면 관리하기가 힘들어질 것이라 생각했습니다.

이것을 해결해보려고, Redux Thunk에 UI 로직 or UI 관련 Dispatch를 넣어볼까 했었으나..
해당 thunk 함수의 재사용성이 매우 떨어질꺼라 생각했고

지난 과정에서는 thunk에서 Promise를 반환해서 await/async를 달아준다면
최신 상태를 가져오지 않을까? 와 같이 생각하고 무작정 시도했었습니다.
(하지만 결과는 리액트의 렌더링 방식으로 인해 제 의도대로 작동되지 않았고... 1단계와 같은 결과가 나왔습니다 🥸 )

여러 시도를 해보는 과정 중!

```javascript
// useDispatchEvent 커스텀 훅
const dispatchEvent = async ({ action, onStateUpdated = () => {} }) => {
  await dispatch(action);
  onStateUpdated(store.getState());
};

// 컴포넌트 내
const handle = () => {
  dispatchEvent({
    action: cartThunk.updateItem(id, { quantity }),
    onStateUpdated: ({ cart }) => {
      !cart.curdAsyncState.isLoaded &&
        alert('서버 오류로 인해 상품 정보 갱신에 실패하였습니다.');
    },
  });
}
```

이번 단계에서 위와 같이 [useDispatchEvent 커스텀 훅](https://github.com/compy-ryu/react-shopping-cart-prod/blob/compy-ryu-step2/src/hooks/useDispatchEvent.js) 내에서 dispatchEvent란 함수를 작성하여
thunk에서 Promise를 반환하고, dispatch에 async / await을 달아준 뒤

useStore 리덕스 훅을 통해 최신 상태를! 핸들러 내에서 꺼내오는 것에 성공하여
현재와 같이 코드를 작성하게 되었습니다.

하지만 반대로 코드 뎁스가 깊어지고
추상화 되어 이해가 되지 않는다면 확인해야할 코드가 늘어날 수도 있고
거쳐가는 레이어만 깊어지고 코드라인만 길어지는게 아닐까? 란 생각도 듭니다.

현업에서는 만약! 비동기 처리를 리덕스에서 한다하였을 때
useEffect로 처리하나요!? 아니면 더 나은 개선법이 있지 않을까 해서 키워드라도 여쭤보고 싶습니다!

glassyi replied:
> 현업에서는 만약! 비동기 처리를 리덕스에서 한다하였을 때
> useEffect로 처리하나요!?

보통은 비동기 처리를 미들웨어에서 해주기 때문에 data fetching 후 추가 처리도 redux-thunk에서 해주는데 화면에 반영하기 위한 상태 업데이트 처리인거고 (데이터, 에러, 로딩 상태),
여기서 콤피가 말하는 비동기처리가 정확히 뭘 이야기 하는걸까요? data fetching 후 ui 추가 처리(alert)를 의미하나요?
변경된 전역 상태에 따라 어떤 ui 처리가 필요하다면, 리듀서에서 상태가 업데이트 된 시점에 useSelector로 리덕스의 변경된 상태값을 반환하여 컴포넌트에 반영되고 그 값에 따라 ui의 변경이 일어나도록 하는 편이에요.

compy-ryu replied:
넵! 맞습니다!
`dispatch` 이후 UI 추가 처리를 여쭤보았었습니다!

만약 `dispatch` 이후, `alert` / `snackbar` / `modal` 등 여러 UI가 결과에 따라 출력되어야 하는 상황이라면
`useSelector`로 상태를 꺼내온 후, 핸들러 바깥 영역에서 `useEffect`로 계속 묶어주는 행위가 코드 가독성을 좋지 않게 만들지 않을까? 라는 뇌피셜이 발동되어서..

위와 같이 핸들러 내부에서도 `dispatch` 이후에 변경된 최신 상태를 가져올 수 있도록 구현해보았었습니다.
(솔직하게 말해서는 별 차이 없는 것 같기도 하고, 오히려 코드 라인이 늘어나 더 읽기 힘들어지지 않았나 생각이 들기도 해요... 😂)

또, 현재 미션에서는 적용되진 않지만!

화면 출력에는 활용하지 않고 단순히 결과 처리에 쓰는 상태까지 `useSelector`로 가져온다면
컴포넌트가 리덕스 내 해당 상태를 구독하고, 값 변경 시 리렌더링 시도 되기에

위와 같은 방식으로 `useStore`으로 값을 꺼낸 뒤, 리렌더링이 일어나지 않도록 구현했었습니다!

compy-ryu commented on src/api/members.js:
```diff
@@ -78,10 +78,27 @@ const requestCheckUserId = (userId) =>
     method: 'GET',
   });
+
+const requestPasswordConfirm = (password) =>
+  request(
+    '/auth/customers/match/password',
+    {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({ password }),
+    },
+    {
+      isAccessTokenUsed: true,
+    },
+  );
+
```
## 시도 3번
### API 요청 함수를 도메인 별로 분리 🗂
API의 경우 서비스가 확장될 시, 여러 영역에서 재사용 되는 케이스가 생길 수도 있지 않을까! 생각했습니다.

그래서 현재와 같이 API 폴더를 만들고, API 요청만 함수들을
도메인 별로 정리해보았는데요.

하지만 이번에 백엔드와 협업했을 때

이 API 함수들을 통해 어떻게, 무슨 값을 보내는지는 바로 알 수 있었지만
**어떤 값들이 오는지**는 이 코드 상에서는 알아볼 수가 없어 불편했었습니다.
(API 명세 문서들을 정리하긴 했지만, 이것도 왔다갔다 보면서 작성하기 쉽지 않더군요..!)

또! 이번에 백엔드와 협업하면서 이 코드를 보여드리니
어떻게 요청 가는지 바로 알 수 있었어서, 많이 좋아하시더라구요.


그리하여, 이곳에서 서버 요청 후 돌아오는 응답 결과도 구조분해한 후
다시 응답 값들을 코드 상에서 바로 알아볼 수 있게 return을 해주는 것도 명시적이고, 협업하기 좋은 코드가 될 것 같기도 한데
이런 접근법도 괜찮을까요?

현업에서는 프로젝트에서 사용하는 API들을 프로젝트 내에서 어떻게 정리하는지 여쭤보아도 될까요!?

glassyi replied:
잘 하신것 같아요! 이것도 회사마다 조금씩 다른데 콤피가 작성한것 처럼 많이 합니다! 여기서 url을 별도로 또 상수로 빼서 관리하기도 하고요~

compy-ryu replied:
엇, 혹시 분리된 API 함수에서 URL을 또 상수로 나눠두는 이유는 무엇인지 여쭤보아도 될까요!?
API 함수에서 다시 한번 API Url을 별도로 상수로 나누어서 관리한다면 얻는 이점이 무엇인지 잘 와닿지가 않습니다! 🤯

compy-ryu commented:
추가로

* 율리님께서 1단계에서 말해주신 인증/인가 부분도 라우터 측에서 바로 알아보고, 처리할 수 있도록 전체적으로 개선해보았습니다.
  * https://github.com/compy-ryu/react-shopping-cart-prod/blob/compy-ryu-step2/src/Routes.jsx

* 회원 탈퇴 기능은 회원 정보 수정 페이지 하단에 작은 글씨로 숨어있습니다!

![스크린샷 2022-06-09 오후 11 14 27](https://user-images.githubusercontent.com/94832610/172868490-a29dca2d-4988-4468-bcb8-efe09f304829.png)

glassyi commented on src/hooks/useFormValidation.js:
```diff
@@ -19,7 +21,22 @@ function useFormValidation(validationList) {
   const isAllPassed = formItems.every((name) => errorList[name] === '');
 
-  return { errorList, isAllPassed, validationForm };
+  const onBlurTextField = async ({ target }) => {
+    if (target.tagName === 'BUTTON') return;
+
+    const formData = getFormData(target.form);
+
+    validationForm({ name: target.name }, formData);
+  };
+
+  const onSubmitForm = (submitAction) => (event) => {
+    event.preventDefault();
+
+    const formData = getFormData(event.target);
+    submitAction(formData);
+  };
```
onSubmitForm를  useFormValidation 훅으로 위치시킨 이유가 뭘까요? 🧐
onBlurTextField는 내부에서 validationForm을 쓰고 있어서 이해가 되는데, onSubmitForm은 단순히 form 핸들러 함수를 여기로 위치만 옮겨 놓았다고 느껴지네요. 
오히려 훅에 있다보니 이 함수가 쓰인 컴포넌트에서 한번에 파악이 안돼요. 게다가 콜백함수를 받아서 처리하고 있어서 더 그렇습니다. 🥲 
```js
 onSubmit={onSubmitForm(fetchWithDrawal)}
```
LoginPage는 onSubmitForm을 쓰지 않았던데 뭐가 더 가독성이 좋은지 한번 다시 확인해 보세요.

compy-ryu replied:
해당 피드백 반영 도중 버그를 찾아, 일단 훅 자체를 다시 개선해보았습니다.. 

브라우저에서 입력란을 자동 완성할 때 (예시: 로그인 페이지)
`onBlur` 이벤트를 사용하다보니, 유효성 검증이 일어나지 않고 그대로 서버에 요청하고 있었습니다 😵‍💫

그래서 번거롭게 해서 죄송하지만 ㅠㅠ 다시 [useForm](https://github.com/compy-ryu/react-shopping-cart-prod/blob/compy-ryu-step2/src/hooks/useForm.js)이란 훅을 작성해보았는데요.
기존과의 차이점 아래와 같습니다.

* `useRef`를 활용하여, Input Value를 Hook에 저장하고, 유효성 검증 시 해당 값을 활용합니다.
(의도: 브라우저에서 자동 완성 시 `onBlur` 이벤트가 발생되지 않아, `onChange` 이벤트에서 값을 가져오기 위해 위와 같이 변경하게 되었습니다.)
* `onSubmit` 이벤트 시 유효성 검증이 이뤄지지 않은 Input에 한해 유효성 검증을 시도하고
오류가 발견될 시 `onSubmit` 이벤트를 중지합니다.

가능한 `onSubmitForm`을 컴포넌트 측으로 이동해보려 했으나..
변경된 구조에서는 더 쉽지 않을 것 같아보여..

일단은 조금은 더 친숙한 형태로 `handle~~`과 같이 컴포넌트에서 핸들러를 전달하고
`useForm` 훅에서는 핸들러로 `event` 객체를 넘겨주는 구조로 수정하였습니다.

**위와 같이 변경한 뒤 개선된 부분은 아래와 같습니다.**
* 브라우저 자동 완성 시 유효성 검증이 일어나지 않는 버그 수정
* 입력 값이 유효성 검증에 통과하지 않을 시 API 요청이 일어나지 않도록 개선
(예: 로그인 시도 시 이메일을 입력하지 않았을 때 API에 요청하지 않고, 즉시 오류 출력)

---

사실 제어 컴포넌트로 바꾸면 가독성도 챙기고, 정말! 쉽게 해결될 일이지만...

이번 단계에서 비제어 컴포넌트와 한번 친해져보고 싶은 욕심으로 인해
계속 이상한 짓(?)을 시도하게 된 부분 사죄드립니다.. 꺼흑 🙇‍♂️

compy-ryu replied:
### 추가 개선 🥸
onSubmitForm으로 핸들러 함수를 받는 상황도 어색하고
JSX에서 함수를 또 감싸고 있다보니 가독성을 해치는게 아닐까!? 란 생각도 들어서

일단 로그인 페이지에서만 아래와 같이 한번 사용 방식을 변경해보았습니다!

```jsx
// 기존 : JSX에서 onSubmitForm으로 핸들러 함수를 래핑.
<form onSubmit={onSubmitForm(handleLogin)} />


// 변경 후 : 함수를 선언할 때 createFormSubmitEvent로 래핑 후, JSX 영역에서 그대로 사용.
const onSubmitLogin = createFormSubmitEvent((event) => {
   // 실행될 내용
});

<form onSubmit={onSubmitLogin} />
```

관련 커밋
* 3f31c445eeca1e84eba7de2a07ce23a9900a3fa8

glassyi commented on src/lib/redux-template/index.js:
```diff
@@ -0,0 +1,23 @@
+  pending: () => ({ isLoading: true, isLoaded: false, error: null }),
+  success: () => ({ isLoading: false, isLoaded: true, error: null }),
+  error: (error) => ({ isLoading: false, isLoaded: true, error }),
+};
```
isLoading 과 isLoaded를 같이 쓸 이유가 있나요? isLoading과 isLoaded는 늘 반대의 상태일텐데 isLoading 한가지만으로 조건을 주기에 충분해 보여서요.

compy-ryu replied:
죄송합니다.. 제가 비동기 상태 관리를 잘못하고 있었습니다 😱
에러 상태일 때 isLoaded 상태가 false가 되어야 하는데 true로 넣어두었네요.

의도는 가끔씩 로딩이 일어나기 전 호출되는 함수에서 활용하기 위해
isLoaded 상태를 만들어주었습니다.

예시
```jsx
if(isLoaded === false && !error) {
    return <p>컨텐츠가 로드되기 전 & 오류가 없을 시, 로딩이 시작되지 않아도 이 내용이 출력됩니다.</p>
}
```

* 8f2f19dc11881aea6cf1f02cddbd5c23c524f85c

glassyi commented on src/pages/CartListPage/Containers/CartItemList/index.jsx:
```diff
@@ -1,38 +1,41 @@
+      action: cartThunk.updateItem(id, { quantity }),
+      onStateUpdated: ({ cart }) => {
+        !cart.curdAsyncState.isLoaded && alert('서버 오류로 인해 상품 정보 갱신에 실패하였습니다.');
+      },
```
!cart.curdAsyncState.isLoaded 가 '서버 오류'에 대한 error plag가 될 수 있나요?

compy-ryu replied:
아뇨... 없습니다 😨
error 상태를 따라가도록 변경하고, alert에서도 에러 메시지를 활용하도록 개선하였습니다!

* 8f2f19dc11881aea6cf1f02cddbd5c23c524f85c

glassyi reviewed (changes requested):
콤피. 2단계까지 고생 많으셨어요~!
이전 단계에서 피드백 드린 내용 확인했고, 질문에 대한 피드백과 몇가지 코드 관련해서 질문 남겼어요!
확인 부탁드릴게요. 🙂 

추가 수정 사항은
로그인에서 로그인 정보 필드에 값이 없다면 아예 login api를 요청하지도 않게 처리하는게 좋을 것 같아요. 지금은 계속 호출하고 있네요. 
![스크린샷 2022-06-11 오후 10 56 47](https://user-images.githubusercontent.com/12928300/173202604-b67c249b-5903-4575-9472-418bac453d4e.png)
.

compy-ryu committed:
🐛 : 잘못된 비동기 상태 관리에 따른 버그 수정

compy-ryu committed:
♻️ : 백엔드 API 선택 관련 구조 개선

- 관련 유틸리티 네이밍 수정
- 선택된 API 기본값 환경변수를 따르도록 변경

compy-ryu committed:
✨ : useForm 커스텀 훅 추가

compy-ryu committed:
✨ : 유효성 검사 유틸리티, 아이디 중복 체크 예외 설정 추가

compy-ryu committed:
♻️ : 로그인 시 유효성 검사에 통과하지 못할 시 서버에 로그인 요청하지 않도록 개선

compy-ryu committed:
🐛 : useForm 변경된 변수 이름 미적용으로 인한 버그 수정

compy-ryu committed:
♻️ : 프로필 편집 페이지 useForm 커스텀 훅으로 교체

compy-ryu committed:
♻️ : 회원가입 페이지 useForm 커스텀 훅으로 교체

compy-ryu committed:
♻️ : 프로필 편집 전 비밀번호 확인 useForm 커스텀 훅 적용

compy-ryu committed:
♻️ : RequireAuth 접근 권한 페이지 사용 시에만 로딩이 출력되도록 개선

compy-ryu commented:
안녕하세요! 율리님!
남겨주신 귀한 피드백들 전반적으로 확인하고 반영해보았습니다!!
(남겨주신 댓글들의 답글으로 반영 내용들 정리하여 남겨두었습니다!)

바쁘신데 자꾸 이상한 코드로 괴롭히는 것 같아
_몬가 몬가_ 마음이 아픕니다 🥸

![스크린샷 2022-06-14 오전 2 24 37](https://user-images.githubusercontent.com/94832610/173410136-b551b74b-6458-4519-b207-50914e3c7b21.png)


그런 의미에서... 리뷰 천천히 리뷰 주셔도 괜찮으니
편하실 때 리뷰 남겨주세요!!

매번 DM으로 신경 써주셔서 감사드려요! 🙇‍♂️

---
아, 그리고! 백엔드 API 선택 부분도 버그 수정하였습니다!

기존과 같이 `yarn start` 로 실행하시면 msw 연결을 기본 상태로 적용되며
`yarn start:jojogreen` 등과 같이 실행하시면 바로 백엔드 API로 연결되도록 버그 수정했습니다!

compy-ryu committed:
🐛 : useFetch 커스텀 훅에서 잘못된 비동기 상태 수정

compy-ryu committed:
♻️ : onSubmit 사용 방식 추가 개선

compy-ryu committed:
♻️ : 회원 탈퇴 페이지 개선

- useForm 커스텀 훅 적용
- 비밀번호가 틀렸을 시 세부 규칙을 보여주지 않도록 개선
- 탈퇴 버튼을 눌렀을 때만, 유효성 검증이 진행되도록 변경

compy-ryu committed:
🐛 : 로그인 실패 이후 다른 페이지로 이동하여도 오류가 계속 유지되는 문제 해결

glassyi reviewed (approved):
콤피 안녕하세요~!
useForm 으로 개선한 내용과 버그들 수정한 부분들 확인했어요~
지금도 많이 개선해 주셨지만 제가 하나하나 작은 개선점들을 찾기보다 
콤피님이 스스로 여러 시도를 해보면서 판단해 가는 게 더 유의미 하지 않을까 해서 
여기서 마무리하려고 합니다!
방학 잘 보내고 있을 텐데 충전 잘 하고 남은 레벨에서도 화이팅해요! 
고생하셨습니다! 🙂
